### PR TITLE
make distclean improvements

### PR DESCRIPTION
**Pull request by shindere**
This PR contains two things:

1. A commit to remove a forgotten configure-generated file when
doing distclean

2. ~A commit to distclean with git clean, when in a git repository.~

### CODE REVIEW

**Comment by gasche** [comment]
I'm not fond of having `dist-clean` use `git clean`, because I often leave notes (or benchmark results, etc.) around in my git repositories, and I use `make dist-clean` when I want to make sure that I want to get back into a reproducible state (removes everything that matters for the build system), and I don't want it to remove my own side-stuff.

**Comment by shindere** [comment]
Ah okay, no problem.

Shall I get rid of the second commit then and just keep the first one?

**Comment by damiendoligez** [comment]
>Shall I get rid of the second commit then and just keep the first one?

Please, yes.

**Comment by mshinwell** [comment]
Well, workflows notwithstanding, `git clean` is actually going to be better than using make rules for cleaning isn't it?  It _ensures_ that everything is back in the correct state.  The very existence of the first part of this patch would seem to prove the point.

I know this can be inconvenient for the reason @gasche proposes, and I have been caught out myself by this recently, but I'm not sure that's a strong argument.

**Comment by gasche** [comment]
If I want to `git clean`, I can run the command and understand the consequences.

**Comment by trefis** [comment]
FTR I completely agree with both of @gasche comments.

**Comment by mshinwell** [comment]
Why does `make clean` not do everything that `make distclean` does?  If that were the case, the latter could be removed, and `git clean` be used instead.  (I think when I do `make clean` I typically reconfigure anyway, even if it isn't required.)

**Comment by trefis** [comment]
> I think when I do make clean I typically reconfigure anyway, even if it isn't required

I don't.

**Comment by lpw25** [comment]
> Why does make clean not do everything that make distclean does?

I think some bootstrap sequences require `make clean` but wouldn't work with `make distclean`, so you'd still need to have a target that does what `clean` does now anyway.

**Comment by shindere** [comment]
Damien Doligez (2018/05/17 07:01 -0700):
> >Shall I get rid of the second commit then and just keep the first one?
> 
> Please, yes.

Done

**Comment by shindere** [comment]
Mark Shinwell (2018/05/17 14:04 +0000):
> Well, workflows notwithstanding, `git clean` is actually going to be better than using make rules for cleaning isn't it?  It _ensures_ that everything is back in the correct state.  The very existence of the first part of this patch would seem to prove the point.
> 
> I know this can be inconvenient for the reason @gasche proposes, and I
> have been caught out myself by this recently, but I'm not sure that's
> a strong argument.

Well perhaps it's better to let people chooose explicitly. Ifthey really
want to cleanup they always have the possibility to call git clean
explicitly. To be honest, my concern is that I do want to git clean and
still keep forgetting the options, that's why I wanted to have them
"scripted" somewhere.

**Comment by shindere** [comment]
Mark Shinwell (2018/05/17 07:34 -0700):
> Why does `make clean` not do everything that `make distclean` does?
> If that were the case, the latter could be removed, and `git clean` be
> used instead.  (I think when I do `make clean` I typically reconfigure
> anyway, even if it isn't required.)

I think the rule of thumb is that makeclean removes
theresultofcompilation,while make distclean also removes the results of
configure.

**Comment by shindere** [comment]
Leo White (2018/05/17 07:47 -0700):
> > Why does make clean not do everything that make distclean does?
> I think some bootstrap sequences require `make clean` but wouldn't
> work with `make distclean`, so you'd still need to have a target that
> does what `clean` does now anyway.

No it's partialclean which is used during bootstrap, I think.

**Comment by lpw25** [comment]
The recipe in the Makefile for a "hard bootstrap" uses `make clean` and its what I've always used when doing one.

**Comment by shindere** [comment]
Leo White (2018/05/17 09:06 -0700):
> The recipe in the Makefile for a "hard bootstrap" uses `make clean`
> and its what I've always used when doing one.

Oh ok

**Comment by gasche** [comment]
> To be honest, my concern is that I do want to git clean and still keep forgetting the options, that's why I wanted to have them "scripted" somewhere.

git supports aliases. You can have an `[alias]` section in your ~/.gitconfig with the following content:

```
[alias]
	hard-clean = clean -xd
```

and then use `git hard-clean`.

**Comment by shindere** [comment]
Sure I can use an alias, it's a good idea, thanks.

So can this PR now be merged?

File: Makefile [hunk 1]
```diff
@@ -1351,6 +1351,7 @@ distclean: clean
 	rm -f boot/ocamlrun boot/ocamlrun$(EXE) boot/camlheader \
 	      boot/ocamlyacc boot/*.cm* boot/libcamlrun.$(A)
 	rm -f config/Makefile byterun/caml/m.h byterun/caml/s.h
+	rm -f config/auto-aux/hashbang4
```
Comment (gasche) [inline Makefile]: I think you could remove the file in `configure` directly, right after HASHBANGSCRIPTS is set.

**Comment by dra27** [comment]
(Just FTR, I've edited the original PR to cross out the second point, as the commit is no longer here)

**Comment by shindere** [comment]
David Allsopp (2018/05/17 23:58 -0700):
> (Just FTR, I've edited the original PR to cross out the second point,
> as the commit is no longer here)

Thanks a lot, @dra27!

**Comment by shindere** [comment]
Gabriel Scherer (2018/05/17 23:57 -0700):
> I think you could remove the file in `configure` directly, right after
> HASHBANGSCRIPTS is set.

Indeed, thanks. Done.

**Review by gasche — APPROVED** [review]
Good to merge if CI passes.

**Comment by shindere** [comment]
Gabriel Scherer (2018/05/18 00:25 -0700):
> gasche approved this pull request.

Thanks!

> Good to merge if CI passes.

I guess we just wait for Travis here.

**Comment by shindere** [comment]
Gabriel Scherer (2018/05/18 01:56 -0700):
> Merged #1784.

Oh, thanks, @gasche!

**Pull request merged**